pkg/controller/completed_workflows.go

```diff
@@ -73,15 +73,25 @@ func CalculateHoursToDelete(retentionPeriodHours, currentHourOfDay int) []string
 	return hoursToDelete
 }
 
+//Calculates a list of all the hours that should be kept given the current time and the retentionperiod in hours
+func CalculateHoursToKeep(retentionPeriodHours int, currentTime time.Time) []string {
+	hoursToKeep := make([]string, 0, retentionPeriodHours+1)
+	for i := 0; i <= retentionPeriodHours; i++ {
+		hoursToKeep = append(hoursToKeep, strconv.Itoa(currentTime.Day())+"."+strconv.Itoa(currentTime.Hour()))
+		currentTime = currentTime.Add(-1 * time.Hour)
+	}
+	return hoursToKeep
+}
+
 // Creates a new selector that selects all completed workflows and workflows with completed hour label outside of the
 // retention window
 func CompletedWorkflowsSelectorOutsideRetentionPeriod(retentionPeriodHours int, currentTime time.Time) *v1.LabelSelector {
-	hoursToDelete := CalculateHoursToDelete(retentionPeriodHours, currentTime.Hour())
+	hoursToKeep := CalculateHoursToKeep(retentionPeriodHours, currentTime)
 	s := CompletedWorkflowsLabelSelector()
 	s.MatchExpressions = append(s.MatchExpressions, v1.LabelSelectorRequirement{
 		Key:      hourOfDayCompletedKey,
-		Operator: v1.LabelSelectorOpIn,
-		Values:   hoursToDelete,
+		Operator: v1.LabelSelectorOpNotIn,
+		Values:   hoursToKeep,
```

Comment:
I wrote up a comment that got lost...

I've a question about migration. What happens when this version of propeller is installed on an existing cluster? It seems to me that it'll start deleting all completed workflows regardless of the "HoursToKeep" because non of them will match the new labels... Can you think of a way to not break that?

Reply:
hi @EngHabu, thanks for your review.
Maybe we can keep the old label selector temporary to not break that. But in this case, the workflows that are supposed to leave but have not will still be stuck

Reply:
done

Reply:
How did you deal with the case you described? The old stuck workflows?

Reply:
The new commits only solve the problem of deleting all completed workflows when this propeller version is installed. The old stuck workflows can be deleted by reducing GCInterval. It will not be a problem if GCInterval is 30 mins.